library.c

```diff
@@ -594,6 +708,106 @@ union resparg {
     double dval;
 };
 
+static zend_string *redis_hash_auth(zend_string *user, zend_string *pass) {
+    zend_string *algo, *digest, *hex;
+    smart_str salted = {0};
+    const php_hash_ops *ops;
+    void *ctx;
+
+    /* No op if there is not username/password */
+    if (user == NULL && pass == NULL)
+        return NULL;
+
+    /* Theoretically inpossible but check anyway */
+    algo = zend_string_init("sha256", sizeof("sha256") - 1, 0);
+    if ((ops = redis_hash_fetch_ops(algo)) == NULL) {
+        zend_string_release(algo);
+        return NULL;
+    }
+
+    /* Hash username + password with our salt global */
+    smart_str_alloc(&salted, 256, 0);
+    if (user) smart_str_append_ex(&salted, user, 0);
+    if (pass) smart_str_append_ex(&salted, pass, 0);
+    smart_str_appendl_ex(&salted, REDIS_G(salt), sizeof(REDIS_G(salt)), 0);
+
+    ctx = emalloc(ops->context_size);
+    ops->hash_init(ctx);
+    ops->hash_update(ctx, (const unsigned char *)ZSTR_VAL(salted.s), ZSTR_LEN(salted.s));
+
+    digest = zend_string_alloc(ops->digest_size, 0);
```

Comment:
You use `digest` as `char *` so zend_string is unnecessary and `emalloc` should be fine